File: src/main/java/xyz/phanta/cbtweaker/multiblock/MultiBlockControllerTileEntity.java
package xyz.phanta.cbtweaker.multiblock;

import io.github.phantamanta44.libnine.tile.L9TileEntityTicking;
import io.github.phantamanta44.libnine.tile.RegisterTile;
import io.github.phantamanta44.libnine.util.data.serialization.AutoSerialize;
import xyz.phanta.cbtweaker.CbtMod;
import xyz.phanta.cbtweaker.common.ActivatableMachine;
import xyz.phanta.cbtweaker.common.MachineDataHolder;
import xyz.phanta.cbtweaker.gui.inventory.UiElement;
import xyz.phanta.cbtweaker.util.block.RedstoneControlHandler;

@RegisterTile(CbtMod.MOD_ID)
public class MultiBlockControllerTileEntity extends L9TileEntityTicking implements ActivatableMachine {

    @AutoSerialize
    private final MachineDataHolder<MultiBlockData<?, ?, ?, ?, ?>> dataHolder = new MachineDataHolder<>(
            () -> MultiBlockData.construct(this,
                    ((MultiBlockControllerBlock)getWorld().getBlockState(getPos()).getBlock()).getMultiBlockType()));

    public MultiBlockControllerTileEntity() {
        markRequiresSync();
        setInitialized();
    }

    public MultiBlockType<?, ?, ?> getMultiBlockType() {
        return dataHolder.getData().getMultiBlockType();
    }

    public boolean isAssembled() {
        return dataHolder.getData().isAssembled();
    }

    @Override
    public boolean isActive() {
        return dataHolder.getData().isActive();
    }

    public RedstoneControlHandler getRedstoneHandler() {
        return dataHolder.getData().getRedstoneHandler();
    }

    public UiElement createRecipeLogicUiElement() {
        return dataHolder.getData().createRecipeLogicUiElement();
    }

    @Override
    public void invalidate() {
        super.invalidate();
        dataHolder.getData().onInvalidated();
    }

    public void notifyHatchChanged(boolean compsDirty) {
        dataHolder.getData().notifyHatchChanged(compsDirty);
    }

    @Override
    public void setDirty() {
        super.setDirty();
    }

    @Override
    protected void tick() {
        dataHolder.getData().tick();
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/multiblock/MultiBlockControllerTileEntity.kt
package st.evening.mc.cbtweaker.multiblock

import io.netty.buffer.Unpooled
import net.minecraft.block.Block
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.network.PacketBuffer
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.ITickable
import net.minecraft.util.math.BlockPos
import st.evening.mc.cbtweaker.CbTweaker
import st.evening.mc.cbtweaker.common.LazyTileEntity
import st.evening.mc.cbtweaker.common.MachineTileEntity
import st.evening.mc.cbtweaker.gui.inventory.UiElement
import st.evening.mc.cbtweaker.network.S2CBindMultiBlockAssembly
import st.evening.mc.cbtweaker.util.component.RedstoneControlHandler
import st.evening.mc.prelude.api.util.game.ClientSide
import st.evening.mc.prelude.api.util.game.ServerSide
import st.evening.mc.prelude.api.util.world.onServer

class MultiBlockControllerTileEntity : LazyTileEntity<MultiBlockData<*>>(), MachineTileEntity, ITickable {
    override fun initData(): MultiBlockData<*> =
        MultiBlockData(this, (world.getBlockState(pos).block as MultiBlockControllerBlock).mbType)

    val mbType: MultiBlockType<*>
        get() = data.mbType

    val assembled: Boolean
        get() = data.assembled

    val assemblyStateClock: Int
        get() = data.assemblyStateClock

    override val isActive: Boolean
        get() = data.isActive

    override val rsHandler: RedstoneControlHandler?
        get() = data.rsHandler

    override fun invalidate() {
        super.invalidate()
        data.onInvalidated()
    }

    fun notifyHatchChanged(compsDirty: Boolean) {
        data.notifyHatchChanged(compsDirty)
    }

    fun onAssemblyChanged(assembly: MultiBlockAssembly<*>?) {
        if (assembly != null) {
            val syncProxy = assembly.syncProxy
            if (syncProxy != null) {
                initSync(syncProxy)
                world.onServer {
                    val syncMgr = syncProxy.syncManager.getServer()
                    syncProxy.dispatchSync(
                        CbTweaker.defns.s2cBindMultiBlockAssembly,
                        S2CBindMultiBlockAssembly(
                            pos,
                            syncMgr.hostId,
                            PacketBuffer(Unpooled.buffer()).also { syncMgr.writeFullState(it) }
                        )
                    )
                }
                return
            }
        }
        initSync(null)
    }

    override fun update() {
        data.tick()
    }

    fun handleInteraction(
        state: IBlockState, player: EntityPlayer, hand: EnumHand,
        face: EnumFacing, hitX: Float, hitY: Float, hitZ: Float
    ): Boolean = data.handleInteraction(state, player, hand, face, hitX, hitY, hitZ)

    @ServerSide
    fun handleBlockUpdate(state: IBlockState, fromBlock: Block, fromPos: BlockPos) {
        data.handleBlockUpdate(state, fromBlock, fromPos)
    }

    @ServerSide
    fun handleDestruction(state: IBlockState) {
        data.handleDestruction(state)
    }

    @ClientSide
    fun bindSync(hostId: Int, syncData: PacketBuffer) {
        data.bindSync(hostId, syncData)
    }

    fun createMachineUiElement(): UiElement? = data.createMachineUiElement()
}
